File: content/tracks/dsa/problems/lc-0049-group-anagrams/solution.py
from typing import Dict, List, Tuple


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        # Anagrams share the same letter counts, so the 26 counts are the group key.
        groups: Dict[Tuple[int, ...], List[str]] = {}
        for word in strs:
            counts = [0] * 26
            for ch in word:
                counts[ord(ch) - ord("a")] += 1
            groups.setdefault(tuple(counts), []).append(word)
        return list(groups.values())

File: content/tracks/dsa/problems/lc-0049-group-anagrams/solution.go
package main

func groupAnagrams(strs []string) [][]string {
	// Anagrams share the same letter counts, so the 26 counts are the group key.
	groups := map[[26]int][]string{}
	for _, word := range strs {
		var counts [26]int
		for i := 0; i < len(word); i++ {
			counts[word[i]-'a']++
		}
		groups[counts] = append(groups[counts], word)
	}
	result := make([][]string, 0, len(groups))
	for _, group := range groups {
		result = append(result, group)
	}
	return result
}
